### ISSUE TITLE
FR: website-based/IP-based split tunneling on MacOS

### ISSUE THREAD

**Issue by Juusttiinn** [tailscale/tailscale#13677]
### What are you trying to do?

Add split tunneling feature on MacOS, such that users are able to let some apps/websites connect the network directly (without passing the VPN/exit node)

### How should we solve this?

Add the split tunneling feature on MacOS

### What is the impact of not solving this?

We can't use the split tunneling feature on MacOS. It will cause some inconvenience, for example: Slower speed with a safety connection that doesn't need passthrough the VPN/exit node; Can't use other VPN to handle some of the connection; etc.

### Anything else?

The PIA VPN (Private Internet Access) has already implemented the split tunneling on the newest macOS and it's open source. I believe this is feasible in development and makes sense in use.

**Comment by agottardo** [comment]
Setting expectations: it is extremely unlikely we'll implement this any time soon. Unlike PIA, we try to be good citizens of the platform and use the native macOS APIs for VPNs, which do not provide this level of customization deep inside the operating system.

**Comment by agottardo** [comment]
> I was wondering if the website-based/IP-based split tunneling also can't be implemented

The majority of our customers resolve the need to route specific traffic to certain hostnames/IPs by using app connectors. Excluding certain routes entirely can definitely be implemented, however I cannot provide any timelines here.

**Comment by sickerin** [comment]
Hi @agottardo, would tailscale split tunneling work well with something like openconnect's split tunneling feature (i've previously used it to split tunnel cisco and fortinet a long time ago)? Currently I'm in a situation where I need to route my traffic thru cisco, however, I also want to access my tailnet at the same time if possible.

**Comment by agottardo** [comment]
@sickerin unless you try to use an exit node, Tailscale only sets routes for the `100.64.0.0/10` address space. 

If you can't reach Tailscale devices while Cisco is running, it likely means Cisco is setting the 0.0.0.0/0 route, so _all_ traffic goes to its interface. Can you configure Cisco to exclude the `100.64.0.0/10` route?

**Comment by sickerin** [comment]
I see, let me take a look at that. But another thing is that when I'm on the corporate wifi network, when I turn tailscale on, I also can no longer access the internal websites. If Tailscale only sets routes for `100.64.0.0/10`, then perhaps it's a DNS issue. I took another look again, and I see that the DNS changes when I have tailscale on. So I guess that's the issue, I'll look into whether I can configure my internal DNS server as a backup if it can't find the site.

**Comment by agottardo** [comment]
Yeah, in your case I would recommend turning off Tailscale DNS in the client settings; and then accessing tailnet resources using their IP. It's likely Cisco takes over DNS resolution on the entire system, so the built-in Tailscale forwarder doesn't work properly.

**Comment by 4piu** [comment]
Is there a workaround? I have Tailscale on MacOS and I frequently need to use exit node and TS subnet routing to access remote hosts. However, I also have docker running services on the same Mac. TS intercepts container Internet traffic whenever connected and causing issues. Is there any way to bypass the docker traffic completely?

**Comment by leolovenet** [comment]
For those looking for a workaround on macOS — I built [tailscale-routes](https://github.com/leolovenet/tailscale-routes), an open-source tool that provides IP-based split tunneling for Tailscale exit nodes on macOS.

How it works: it injects more-specific routes (e.g. `/8`, `/16`, `/24`) into the kernel routing table to override Tailscale's `/1` catch-all routes, so designated IP ranges bypass the VPN and route through the local physical gateway.

Key highlights:
- A C helper program uses PF_ROUTE sockets for bulk kernel route operations — handles 7,500 routes in ~150ms
- Python controller daemon auto-detects exit node connect/disconnect and Wi-Fi gateway changes
- Hot-reload: edit the route config file and changes apply within 5 seconds, no restart needed
- Zero external dependencies — only needs Xcode CLT (provides `cc` and `python3`)

This is a userspace workaround until official macOS split tunnel support lands. Tested on macOS Sequoia 15.x.
